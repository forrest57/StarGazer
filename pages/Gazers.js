import { StatusBar } from 'expo-status-bar'
import { View, SafeAreaView, FlatList, Text } from 'react-native'
import tw from 'twrnc'
import { Ionicons } from '@expo/vector-icons'
import { SimpleLineIcons } from '@expo/vector-icons'
import AnimatedButton from '../components/animatedButton'
import GazerCard from '../components/gazerCard'

const chevronLeftIcon = () => (
  <View
    style={tw` bg-gray-700 flex items-center justify-center rounded-md p-4 w-full`}>
    <Ionicons name='chevron-back-sharp' size={25} color='gray' />
  </View>
)
const gearIcon = () => (
  <View
    style={tw` bg-gray-700 flex items-center justify-center rounded-md p-4 w-full `}>
    <SimpleLineIcons name='settings' size={25} color='gray' />
  </View>
)

export default Gazers = ({ navigation, route }) => {
  const cardRenderer = ({ item }) => (
    <GazerCard login={item.login} avatar={item.avatar_url} />
  )

  const goHome = () => navigation.navigate('Home')

  return (
    <SafeAreaView
      style={[
        { marginTop: StatusBar.currentHeight || 0 },
        tw`flex flex-col justify-between bg-gray-800 pt-7 h-full w-full `,
      ]}>
      <View
        style={[
          tw`flex flex-row items-center p-1 w-full  px-5 mb-3 rounded-md`,
          { justifyContent: 'space-between' },
        ]}>
        <AnimatedButton Component={chevronLeftIcon} pressFunction={goHome} />
        <Text style={{ fontSize: 28, color: 'gray' }}>{route.params.repo}</Text>
        <AnimatedButton Component={gearIcon} />
      </View>
      <FlatList
        style={tw`w-full px-5`}
        data={route.params.data}
        renderItem={cardRenderer}
        keyExtractor={(item) => item.id}
      />
      <StatusBar style='light' />
    </SafeAreaView>
  )
}

