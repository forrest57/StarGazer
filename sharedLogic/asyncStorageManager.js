import AsyncStorage from '@react-native-async-storage/async-storage'

export const appendRepoToHistory = async (repoString) => {
  try {
    const lowerCased = repoString.toLowerCase()
    const repoHistory = await getRepoHistory()
    if (!repoHistory.includes(lowerCased)) {
      repoHistory.unshift(lowerCased)
      const stringified = JSON.stringify(repoHistory)
      await AsyncStorage.setItem('RepoHistory', stringified)
      return true
    } else {
      return false
    }
  } catch (e) {
    return false //edge case, does not seem really testable
  }
}

export const getRepoHistory = async () => {
  try {
    const repoHistory = await AsyncStorage.getItem('RepoHistory')
    if (repoHistory !== null) {
      return JSON.parse(repoHistory)
    } else {
      return []
    }
  } catch (_) {
    return failMsg //edge case, again
  }
}
export const clearRepoHistory = async () => {
  const successMsg = ['Success', 'Cleared repo history']
  const failMsg = ['Error', 'There was an error clearing the cache history']

  try {
    const _ = await AsyncStorage.removeItem('RepoHistory')
    return successMsg
  } catch (_) {
    return failMsg //edge case, once more
  }
}
