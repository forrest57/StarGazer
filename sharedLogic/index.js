import AsyncStorage from '@react-native-async-storage/async-storage'
import axios from 'axios'

const baseUrl = 'https://api.github.com/repos/'

export const getGazers = async (user, repo) => {
  const res = await axios.get(`${baseUrl}${user}/${repo}/stargazers`)
  return res
}

export const validateRequest = (user, repo) => {
  if (user && repo) {
    return getGazers(user, repo)
      .then((res) => {
        appendRepoToHistory()
        return res
      })
      .catch((err) => {
        switch (err.response.status) {
          case 404:
            throw ['error', 'No repo found with the provided data']
          case 403:
            throw ['error', 'Api rate exceeded, try again later.']
          default:
            throw ['error', 'Unknown error, try again later']
        }
      })
  } else {
    throw ['error', 'empty search fields, add data first']
  }
}

export const appendRepoToHistory = async (repoString) => {
  try {
    const lowerCased = repoString.toLowerCase()
    const repoHistory = await getRepoHistory()
    if (!repoHistory.includes(lowerCased)) {
      repoHistory.unshift(lowerCased)
      const stringified = JSON.stringify(repoHistory)
      await AsyncStorage.setItem('RepoHistory', stringified)
      return true
    } else {
      return false
    }
  } catch (e) {}
}

export const getRepoHistory = async () => {
  const repoHistory = await AsyncStorage.getItem('RepoHistory')
  if (repoHistory !== null) {
    return JSON.parse(repoHistory)
  } else {
    return []
  }
}
export const clearRepoHistory = async () => {
  const successMsg = ['Success', 'Cleared repo history']
  const failMsg = ['Error', 'There was an error clearing the cache history']

  try {
    const _ = await AsyncStorage.removeItem('RepoHistory')
    return successMsg
  } catch (_) {}
  return failMsg
}
