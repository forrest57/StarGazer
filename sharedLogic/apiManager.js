import axios from 'axios'
import { appendRepoToHistory } from './asyncStorageManager'

const baseUrl = 'https://api.github.com/repos'

export const getGazers = async (user, repo) => {
  const res = await axios.get(`${baseUrl}/${user}/${repo}/stargazers`)
  return res
}

export const validateRequest = async (user, repo) => {
  if (user && repo) {
    try {
      const res = await getGazers(user, repo)
      await appendRepoToHistory(`${user}/${repo}`)
      return res
    } catch (err) {
      switch (err.response.status) {
        case 404:
          throw ['error', 'No repo found with the provided data']
        case 403:
          throw ['error', 'Api rate exceeded, try again later.']
        default:
          throw ['error', 'Unknown error, try again later']
      }
    }
  } else {
    throw ['error', 'empty search fields, add data first']
  }
}
