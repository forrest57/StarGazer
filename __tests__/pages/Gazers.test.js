import {
  cleanup,
  fireEvent,
  render,
  screen,
  waitForElementToBeRemoved,
} from '@testing-library/react-native'

import Gazers from '../../pages/Gazers'
import React from 'react'
import { loadNextPage } from '../../sharedLogic/apiManager'
import renderer from 'react-test-renderer'

afterEach(cleanup)

const route = {
  params: {
    repo: 'loremIpsum',
    data: [
      { id: 1, login: 'lorem', avatar_url: null },
      { id: 2, login: 'lorem', avatar_url: null },
      { id: 3, login: 'lorem', avatar_url: null },
    ],
  },
}
const moreItems = {
  repo: 'loremIpsum',
  data: [
    { id: 4, login: 'new', avatar_url: null },
    { id: 5, login: 'new2', avatar_url: null },
    { id: 6, login: 'new3', avatar_url: null },
  ],
}

jest.mock('../../sharedLogic/apiManager.js', () => ({
  loadNextPage: jest.fn(),
}))

const eventData = {
  nativeEvent: {
    contentOffset: {
      y: -5000,
    },
    contentSize: {
      // Dimensions of the scrollable content
      height: 200,
      width: 100,
    },
    layoutMeasurement: {
      // Dimensions of the device
      height: 100,
      width: 100,
    },
  },
}
const routeEmpty = {
  params: {
    repo: 'loremIpsum',
    data: [],
  },
}
describe('<Gazers />', function () {
  //prettier kept adding a semicolon
  it('renders correctly', () => {
    const tree = renderer.create(<Gazers route={route} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('correctly renders prop', () => {
    const tree = renderer.create(<Gazers route={route} />).toJSON()
    const fullRepoName = tree.children[0].children[1].children[0]
    expect(fullRepoName).toBe(route.params.repo)
  })
  it('correctly renders conditional item', () => {
    render(<Gazers navigation={null} route={routeEmpty} />)
    const conditionalComponent = screen.queryByText('No Stargazers?')
    expect(conditionalComponent).toBeTruthy()
  })
  it('scrolls to bottom and loads new items', async () => {
    render(<Gazers route={route} />)
    expect(() => screen.getByText('new')).toThrow() //not shown
    loadNextPage.mockResolvedValueOnce(moreItems)
    fireEvent.scroll(screen.getByTestId('List'), eventData)
    await waitForElementToBeRemoved(
      async () => await screen.findByText(/loading more.../i),
      {
        timeout: 3000,
      }
    )
    await waitFor(() => {
      expect(getByText(/new/i)).toBeTruthy()
    })
  })
})
