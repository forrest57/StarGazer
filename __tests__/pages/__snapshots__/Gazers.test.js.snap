// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Array [
      Object {
        "marginTop": 0,
      },
      Object {
        "backgroundColor": "#1f2937",
        "display": "flex",
        "flexDirection": "column",
        "height": "100%",
        "justifyContent": "space-between",
        "paddingTop": 28,
        "width": "100%",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderRadius": 6,
          "display": "flex",
          "flexDirection": "row",
          "marginBottom": 12,
          "paddingBottom": 4,
          "paddingLeft": 20,
          "paddingRight": 20,
          "paddingTop": 4,
          "width": "100%",
        },
        Object {
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "display": "flex",
          "marginTop": 8,
          "transform": Array [
            Object {
              "scale": 1,
            },
          ],
          "width": "auto",
        }
      }
    >
      <View
        accessible={true}
        activeOpacity={1}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "borderRadius": 6,
            "display": "flex",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#374151",
              "borderRadius": 6,
              "display": "flex",
              "justifyContent": "center",
              "paddingBottom": 16,
              "paddingLeft": 16,
              "paddingRight": 16,
              "paddingTop": 16,
              "width": "100%",
            }
          }
        >
          <Text />
        </View>
      </View>
    </View>
    <Text
      style={
        Object {
          "color": "gray",
          "fontSize": 28,
        }
      }
    >
      loremIpsum
    </Text>
    <View
      collapsable={false}
      style={
        Object {
          "display": "flex",
          "marginTop": 8,
          "transform": Array [
            Object {
              "scale": 1,
            },
          ],
          "width": "auto",
        }
      }
    >
      <View
        accessible={true}
        activeOpacity={1}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "borderRadius": 6,
            "display": "flex",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#374151",
              "borderRadius": 6,
              "display": "flex",
              "justifyContent": "center",
              "paddingBottom": 16,
              "paddingLeft": 16,
              "paddingRight": 16,
              "paddingTop": 16,
              "width": "100%",
            }
          }
        >
          <Text />
        </View>
      </View>
    </View>
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "avatar_url": null,
          "id": 1,
          "login": "lorem",
        },
        Object {
          "avatar_url": null,
          "id": 2,
          "login": "lorem",
        },
        Object {
          "avatar_url": null,
          "id": 3,
          "login": "lorem",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "paddingLeft": 20,
        "paddingRight": 20,
        "width": "100%",
      }
    }
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#374151",
              "borderRadius": 6,
              "display": "flex",
              "flexDirection": "row",
              "marginBottom": 8,
              "marginTop": 8,
              "paddingBottom": 16,
              "paddingLeft": 16,
              "paddingRight": 16,
              "paddingTop": 16,
            }
          }
        >
          <Image
            source={
              Object {
                "uri": null,
              }
            }
            style={
              Object {
                "aspectRatio": 1,
                "borderRadius": 9999,
              }
            }
          />
          <Text
            style={
              Object {
                "color": "gray",
                "fontSize": 22,
                "marginLeft": 15,
              }
            }
          >
            lorem
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#374151",
              "borderRadius": 6,
              "display": "flex",
              "flexDirection": "row",
              "marginBottom": 8,
              "marginTop": 8,
              "paddingBottom": 16,
              "paddingLeft": 16,
              "paddingRight": 16,
              "paddingTop": 16,
            }
          }
        >
          <Image
            source={
              Object {
                "uri": null,
              }
            }
            style={
              Object {
                "aspectRatio": 1,
                "borderRadius": 9999,
              }
            }
          />
          <Text
            style={
              Object {
                "color": "gray",
                "fontSize": 22,
                "marginLeft": 15,
              }
            }
          >
            lorem
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "#374151",
              "borderRadius": 6,
              "display": "flex",
              "flexDirection": "row",
              "marginBottom": 8,
              "marginTop": 8,
              "paddingBottom": 16,
              "paddingLeft": 16,
              "paddingRight": 16,
              "paddingTop": 16,
            }
          }
        >
          <Image
            source={
              Object {
                "uri": null,
              }
            }
            style={
              Object {
                "aspectRatio": 1,
                "borderRadius": 9999,
              }
            }
          />
          <Text
            style={
              Object {
                "color": "gray",
                "fontSize": 22,
                "marginLeft": 15,
              }
            }
          >
            lorem
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
